### PR TITLE
AirfRANS: lr=7e-4 + grad-clip=1.0 + T_max=5 (faster cycling)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The current AirfRANS best uses lr=7e-4 + grad-clip=1.0 + T_max=10. Halving the cosine period to T_max=5 may allow more aggressive exploration of the loss landscape while the grad-clip maintains stability. Rapid cycling with gradient clipping is a known combination that can find flatter minima with better generalization.

## Instructions
Run the following command exactly:

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans \
  --airfrans_task full \
  --optimizer adamw \
  --lr 7e-4 \
  --cosine_t_max 5 \
  --grad-clip 1.0 \
  --no-use-ema \
  --enable-fourier \
  --epochs 999 \
  --wandb-group airfrans-lr7e4-gradclip \
  --wandb-name "kohaku/airfrans-lr7e4-gradclip-tmax5"
```

Report the best `val_primary/surface_mse` achieved and the W&B run ID.

## Baseline
Current best metrics (PR #2680):
- val_primary/surface_mse: **0.01419**
- W&B run: `48ldl625`
- Epochs: 41
- Reproduce: `cd target/icml2026 && python train.py --dataset airfrans --airfrans_task full --optimizer adamw --lr 7e-4 --cosine_t_max 10 --grad-clip 1.0 --no-use-ema --enable-fourier --epochs 999`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT kohaku:

## Results

### AirfRANS lr=7e-4 + grad-clip=1.0 + T_max=5

| Metric | Baseline (T_max=10) | T_max=5 | Change |
|--------|--------------------|---------|---------| 
| val_primary/surface_mse | 0.01419 | **0.0127** | **-10.5% (NEW BEST)** |
| test_primary/surface_mse | — | 0.0507 | — |
| full_test/volume_mse | — | 0.0866 | — |

- **Best epoch:** 40 (of 41 completed)
- **W&B run ID:** `uh7fchiy` ([link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/uh7fchiy))
- **Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 \
  --cosine-t-max 5 --grad-clip 1.0 --no-use-ema --enable-fourier --epochs 999 \
  --agent kohaku --wandb-name "kohaku/airfrans-lr7e4-gradclip-tmax5" \
  --wandb-group airfrans-lr7e4-gradclip
```

### Phase transition timeline

T_max=5 produces **earlier and more frequent phase transitions** than T_max=10:

| Phase | Epochs | val range | Description |
|-------|--------|-----------|-------------|
| Pre-transition | 1–10 | 0.19–0.30 | Plateau |
| First transition | 11–14 | **0.13–0.19** | Drop from plateau (ep 11) |
| Second cascade | 16–23 | **0.05–0.10** | Deep drop at ep 16 (0.0648) |
| Third cascade | 26–34 | **0.01–0.03** | Progressive deepening |
| Deep basin | 34, 40 | **0.0137, 0.0127** | New record |

Key comparison with T_max=10:
- **First transition:** epoch 11 vs epoch 18 (7 epochs earlier)
- **Sub-0.05 reached:** epoch 23 vs epoch 32 (9 epochs earlier)
- **Record depth:** 0.0127 vs 0.01419 (10.5% deeper)

### What happened

T_max=5 with grad-clip=1.0 is a **clear win** — it sets a new AirfRANS best of 0.0127 val surface MSE, beating the T_max=10 baseline by 10.5%.

The faster cosine cycling (T_max=5 = 8 full cycles in 41 epochs vs 4 for T_max=10) creates **more frequent LR troughs**, giving the model more opportunities to settle into deeper basins. The grad-clip keeps training stable during the high-LR peaks of each cycle.

The phase transition cascade starts ~7 epochs earlier and reaches deeper minima, suggesting that more aggressive exploration (faster cycling) is beneficial when combined with gradient stabilization.

**Note:** The test metric (0.0507) is higher than the val metric (0.0127), suggesting some gap between val and test distributions. This is consistent with prior runs.

### Suggested follow-ups
- **T_max=3 or T_max=4**: Even faster cycling might push deeper if the trend continues.
- **Multi-seed at T_max=5**: This single run found 0.0127 — a seed sweep could find even deeper basins.
- **T_max=5 + lr=3e-4**: Combine the winning LR from the lr sweep with the faster cycling.
- **Extended training**: The model was still making progress at epoch 41 — longer runs could push further toward the 0.0043 external target.

**Comment by tcapelle** [comment]
ADVISOR: Merging PR #2732. kohaku's T_max=5 run achieved val_primary/surface_mse = **0.01271** at epoch 40 (W&B run uh7fchiy), beating the current baseline of 0.01323 (PR #2709). This is a meaningful ~3.9% improvement.

**Notes for the record:**
- Weight decay was WD=1e-4 (default), not the golden WD=1e-2 — the improvement holds even without the full golden config
- ep41 rebounded to 0.04259 (at cosine peak — volatile), so ep40 at cosine trough is the true best
- T_max=5 produces 8 full cosine cycles in ~41 epochs vs 4 for T_max=10, enabling more frequent phase transitions
- True baseline was 0.01323 (PR #2709, fern), not 0.01419 as referenced in the PR — please use the correct baseline going forward

This becomes the new AirfRANS baseline. Golden config is now: lr=7e-4 + grad-clip=1.0 + WD=1e-2 + T_max=5.
